project.c

```diff
@@ -1188,15 +1194,65 @@ void ALU_Control(BIT* ALUOp, BIT* funct, BIT* ALUControl)
   ALUControl[0] = or_gate(slt_and_gate_final,or_and_gate_final);
 }
 
+void adder1(BIT A, BIT B, BIT CarryIn, BIT* CarryOut, BIT* Sum)
+{
+  // TODO: implement a 1-bit adder
+  BIT x0 = xor_gate(A, B);
+  *Sum = xor_gate(CarryIn, x0);
+  
+  BIT y0 = and_gate(x0, CarryIn);
+  BIT y1 = and_gate(A, B);
+  *CarryOut = or_gate(y0, y1);
+}
+
+void ALU1(BIT A, BIT B, BIT CarryIn, BIT Binvert, BIT Less,
+BIT Control0,BIT Control1, BIT Control2,
+BIT* Result, BIT* CarryOut, BIT * Set)
+{
+  BIT x0 = multiplexor2(Binvert, B, not_gate(B));
+
+  BIT y0 = and_gate(A,x0);
+  BIT y1 = or_gate(A,x0);
+  
+  BIT y2 = FALSE;
+  adder1(A, B, CarryIn, CarryOut, &y2);
+  *Set = y2;
+  BIT y3 = Less;
+  *Result = multiplexor8(Control0, Control1, Control2, y0, y1, y2, y3);
+}
 void ALU(BIT* ALUControl, BIT* Input1, BIT* Input2, BIT* Zero, BIT* Result)
 {   
   // TODO: Implement 32-bit ALU
   // Input: 4-bit ALUControl, two 32-bit inputs
   // Output: 32-bit result, and zero flag big
   // Note: Can re-use prior implementations (but need new circuitry for zero)
-  
+  BIT Less = FALSE;
+  BIT Set = FALSE;
+  BIT CarryIn = FALSE;
+  BIT CarryOut = FALSE;
+  BIT s0 = ALUControl[2];
+  BIT s1 = ALUControl[1];
+  BIT s2 = ALUControl[0];
+  BIT Op = multiplexor8(s0,s1,s2,//8 inputs??);
+  ALU1(Input1[0], Input2[0], /*Binvert, CarryIn, CarryOut??*/, Less, 
+    &Op, &Result[0],&Set);
+  for (int i = 1; i < 32; i++){
+    BIT x0 = multiplexor2(Input2_invert[i], Input2[i], not_gate(Input2[i]));
+    BIT y0 = and_gate(Input1[i],x0);
+    BIT y1 = or_gate(Input1[i],x0);
+    BIT y2 = FALSE;
+    adder1(Input1[i], x0, &CarryIn, &CarryOut, &y2);
+    *Set = y2;
+    BIT y3 = Less;
+    ALU1(Input1[i],Input2[i],/*Binvert,CarryIn,CarryOut??*/Less,&Op,&Result[i],&Set);
```

Comment:
for both binvert and carryin, you pass in the bit that only changes for subtraction (second most significant bit from the ALUcontrol). I might be wrong, but I think for carryout when you call the alu, you just create a carryout variable and pass it in like,
BIT Carryout;
ALU1(Input1[0], Input2[0], ALUControl[3], ALUControl[3], Less, 
    ALUControl[0], ALUControl[1], &Result[0], CarryOut, &Set);